File: src/models/Recomposition.ts
export type RecompositionCause =
  | "unstable_lambda"
  | "unstable_parameter"
  | "state_read"
  | "parent_recomp"
  | "collection_change"
  | "unknown";

export interface RecompositionMetrics {
  sinceLastObservation: number;
  sinceLastInteraction: number;
  rolling1sAverage: number;
  total: number;
  skipCount: number;
  durationMs?: number;
}

export interface RecompositionNodeInfo {
  id: string;
  composableName?: string;
  resourceId?: string | null;
  testTag?: string | null;
  total?: number;
  skipCount?: number;
  rolling1sAverage?: number;
  durationMs?: number;
  likelyCause?: RecompositionCause;
  parentChain?: string[];
  stableAnnotated?: boolean;
  rememberedCount?: number;
}

export interface TopRecompositionEntry {
  rank: number;
  composableName?: string;
  resourceId?: string | null;
  recompositionId?: string;
  recompCount: number;
  recompPerSecond: number;
  recompDurationMs?: number;
  likelyCause: RecompositionCause;
  parentChain?: string[];
}

export interface RecompositionSummary {
  totalRecompositions: number;
  averagePerSecond: number;
  averageRecompositionDurationMs?: number;
  topRecompositions: TopRecompositionEntry[];
}

File: android/control-proxy/src/main/kotlin/dev/jasonpearson/automobile/ctrlproxy/models/Recomposition.kt
package dev.jasonpearson.automobile.ctrlproxy.models

import kotlinx.serialization.Serializable

@Serializable
data class RecompositionEntry(
    val id: String,
    val composableName: String? = null,
    val resourceId: String? = null,
    val testTag: String? = null,
    val total: Int = 0,
    val skipCount: Int = 0,
    val rolling1sAverage: Double? = null,
    val durationMs: Double? = null,
    val likelyCause: String? = null,
    val parentChain: List<String>? = null,
    val stableAnnotated: Boolean? = null,
    val rememberedCount: Int? = null,
)

@Serializable
data class RecompositionSnapshot(
    val timestamp: Long,
    val applicationId: String,
    val entries: List<RecompositionEntry>,
)
